### PR TITLE
Windows webauthn wrapper

### PR DESCRIPTION
**Pull request by tobiaszheller**
implementes wrapper needed for https://github.com/gravitational/teleport/pull/16360

What has changed?

support for fido devices and windows hello on windows platform

Support for extension: app_id will be provided in other PR.
Integration/manual tests will be provided in other PR.
Deleting credentials will be added in separate PR (only supported in v4)

Issue: https://github.com/gravitational/teleport/issues/15148

### CODE REVIEW

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [inline lib/auth/winwebauthn/webauthn_windows.go:311]: A Q about Windows APIs: could we simply set the highest version number we support and let older APIs do the best they can? Or does it not work like that?

**Review by codingllama — COMMENTED** [review]
Mostly style Qs, but there a few legitimate comments/Qs in there. Please reply/solve as you address the comments so we don't get lost.

Do you think it's worth trying to write a "happy-path" test for this? I realize it's mostly type conversions, but I think it's best to get some coverage regardless. (If you do please keep in mind that CI is essentially Linux.)

**Comment by tobiaszheller** [comment]
> Do you think it's worth trying to write a "happy-path" test for this? I realize it's mostly type conversions, but I think it's best to get some coverage regardless. (If you do please keep in mind that CI is essentially Linux.)

To be honest I don't see too much value in that test :( I believe that we will get more benefit from self-automated integration tests. Here is WIP that I am working currently on https://github.com/gravitational/teleport/blob/4e222cf005dbc0555df50bf78cd3f6c4a355c9db/lib/auth/winwebauthn/webauthn_windows_integration_test.go. You will need to have window machine and it will require human interactions but I see a lot value in that kind of tests.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:79]: I have done that on purpose (but forgot to add comment for it). I think it will be better to test it when v5 goes out instead of blindly supporting it. What's your opinion on that?

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:49]: it's the API version, i will rename it to API version.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:330]: i am planning to add support for U2fAppId in separate PR but will remove that line and keep only todo comment

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:612]: all those comments in Ctypes are copied from WIndows API documentation. Should we edit them? Or maybe they don't bring any value and we should just provide proper link to header file?

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:311]: I am not sure if I understand. 
I think the problem is that there are multiple kind of "versions" used.

First type is API version (v1-v4).
Other type is version of structure in endpoints. (They don't match API version at all).
https://github.com/microsoft/webauthn/blob/7ab979cc833bfab9a682ed51761309db57f56c8c/webauthn.h#L36-L97

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:48]: I have added following comment:
```
// client keeps diagnostic informations about windows webauthn support.
``` 
but if you have better naming idea, I am open :)

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:129]: non-nill error happens when there is some system issue, while non 0 is when API returns error (for example user cancel dialog)

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:127]: i tyhink here we will get some system error however I have never ended up having error here. I used trace.wrap()

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:325]: I have added following info:
```
		// https://github.com/microsoft/webauthn/blob/7ab979cc833bfab9a682ed51761309db57f56c8c/webauthn.h#L36-L97
		// contains information about different versions.
		dwVersion:                         c.getMakeCredentialOptionsVersion(),
``` 
let me know if this is enough or not

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go:310]: I have checked it and it should work

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: I have added more info. 

> How does it interact with FIDO2
> How does it interact with Hello

This is hidden from user and I have not idea how it interacts, not sure what I need to add here.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: I have added more info.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: Yes, for hello they are always resident, similary to TouchID implementation. I assume that this field is used for FIDO flow but this is shot because we are missing correct docs.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: `nativeImpl` or `nativeWindows`, as we call the interface `native` in the next PR?

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Fair. I would suspect the "err" error rarely happens and most of the time we'd get a non-zero code, but no harm in asking.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: That's helpful, thanks.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Let's mention somewhere in docs and/or comments that Hello keys are always resident, that's good to know.

I have an ongoing update for duo-labs/webauthn, but we never set "preferred" anyway, so feel free to tweak the TODO/code accordingly.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Ack. Remind me to test as well, my API version is likely older than yours. Meanwhile, any chance you could find some documentation that explains how Windows deals with this?

I think this get considerably easier if we can just set the highest version number we support and move on.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Great, thanks.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Ack, sounds good.

File: lib/auth/webauthnwin/ctypes_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/ctypes_windows.go:152]: nit: Use all private fields, so it's consistent with other structs?

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:45]: nit:

```suggestion
	modUser32               = windows.NewLazySystemDLL("user32.dll")
```

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:311]: We have other threads about this, but just so it doesn't lost: let's follow up on whether we need to adjust versions or if we can just always set `6`.

Same for similar logic in other places.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Bump

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Bump.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Looking better, thanks.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Bump.

**Review by codingllama — COMMENTED** [review]
Looking good. I think we have a few small tweaks to go, so I'll wait until it's 100% done to approve. Give me a nudge when it's are ready.

> To be honest I don't see too much value in that test :( I believe that we will get more benefit from self-automated integration tests (...)

Ack. The issue here is that these almost never get run. The plus of automated is that we get as much coverage as we can, plus someone on other OS could do small tweaks with higher confidence.

I would shoot for as much automated, system-independent testing as we can, but I'll leave this up to you.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: added comments and update MakeCredential func docs with info about resident keys.

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: yes, DER added in comment

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/winwebauthn/webauthn_windows.go]: ok i was able to test it on older API and you were right, we can set highest number on older API and it still works, I have changed code to only contain newest version

**Comment by tobiaszheller** [comment]
@codingllama all comments should be fixed now, PTAL

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:101]: nit: Use `*nativeImpl` as receiver consistently?

File: lib/auth/winwebauthn/webauthn_windows.go
Comment (codingllama) [reply lib/auth/winwebauthn/webauthn_windows.go]: Nice.

**Review by codingllama — APPROVED** [review]

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go]: done

**Comment by tobiaszheller** [comment]
@jakule @tcsc it should be ready for review

**Review by codingllama — APPROVED** [review]

File: lib/auth/webauthnwin/ctypes_windows.go
Comment (jakule) [inline lib/auth/webauthnwin/ctypes_windows.go:163]: nit:
```suggestion
	// 0 => encoded as CBOR null value.
```

File: lib/auth/webauthnwin/ctypes_windows.go
Comment (jakule) [inline lib/auth/webauthnwin/ctypes_windows.go:165]: nit:
```suggestion
	// Encoded CBOR attestation information
```

File: lib/auth/webauthnwin/ctypes_windows.go
Comment (jakule) [inline lib/auth/webauthnwin/ctypes_windows.go:324]: nit: I don't think that C casting is needed here.
```suggestion
	pwszVer    *uint16 // "2.0"
```

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (jakule) [inline lib/auth/webauthnwin/webauthn_windows.go:126]: Are you sure that the order here is correct? https://pkg.go.dev/golang.org/x/sys/windows#LazyProc.Call

> Callers must inspect the primary return value to decide whether an error occurred (according to the semantics of the specific function being called) before consulting the error.

From what I understand, `ret` should be checked before `err` here in all other cases.

**Review by jakule — APPROVED** [review]
LGTM, I left a few comments.

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go]: @jakule thanks for comment, i have added better handling of it in https://github.com/gravitational/teleport/pull/16634/commits/48e8b23ea71334935e84705e92379450e78b9cab

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:268]: Could we do handle this like the others that do `if ret != 0 { return trace.Wrap(getErrorNameOrLastErr(...)) }` ?

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:195]: nit:

```suggestion
	ret, _, err := procWebAuthNAuthenticatorMakeCredential.Call(
```

I'm not seeing a reason to name this one error differently?

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [inline lib/auth/webauthnwin/webauthn_windows.go:288]: Ditto:

```suggestion
	ret, _, err := procWebAuthNIsUserVerifyingPlatformAuthenticatorAvailable.Call(
```

**Review by codingllama — APPROVED** [review]

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go:268]: here API does not return `hresult` like in other API calls but returns dword, which is APi version. So in most cases it will be non 0 but this is fine.

More details: https://github.com/microsoft/webauthn/blob/7ab979cc833bfab9a682ed51761309db57f56c8c/webauthn.h#L895-L902

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go:195]: i have named it differently to indicate that this is non standard error. For example you cannot use:
```
if err!=nil {
   return nil
}
```
because if everything went fine inside that value there will be error equal to `This operation finished successful` or something similar. 
> // The returned error is always non-nil, constructed from the result of GetLastError.
// Callers must inspect the primary return value to decide whether an error occurred
// (according to the semantics of the specific function being called) before consulting
// the error. The error always has type syscall.Errno.


I don't have strong preferences and can revert it back to err.

Other option is ignore it completely which  is done in few libraries: https://github.com/AllenDang/w32/blob/ad0a36d80adcd081d5c0dded8e97a009b486d1db/user32.go#L219-L226
i think ignoring it also makes some sense. I have actually never received any error there.
If API fails with something - then we got non 0 error which we translate using another API call.
When something is mixed with input types - it results in panic.

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [reply lib/auth/webauthnwin/webauthn_windows.go:268]: Gotcha, maybe leave a comment? It's hard to spot the difference.

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (codingllama) [reply lib/auth/webauthnwin/webauthn_windows.go:195]: I think it's important to maintain a consistent idiom, regardless of the choice we make. I'm OK ignoring the err if you think that's best - up to you.

Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go:195]: ok, I will keep error checks just in case but will rename it to `err`.

File: lib/auth/webauthnwin/webauthn_windows.go
Comment (tobiaszheller) [reply lib/auth/webauthnwin/webauthn_windows.go:271]: added

**Review by codingllama — APPROVED** [review]

**Pull request merged**
